pages/shared/editHelpers.js

```diff
@@ -0,0 +1,680 @@
+    // If you add to this list, you also need to add aceMode handlers in instructorFileEditor.js
+    const extCanEdit = ['.py', '.html', '.json', '.txt', '.md'];
+    return extCanEdit.includes(path.extname(file));
+}
+
+function canMoveFile(file) {
+    const cannotMove = ['info.json', 'infoAssessment.json', 'infoCourseInstance.json', 'infoCourse.json'];
+    return (! cannotMove.includes(path.basename(file)));
+}
+
+function getFiles(options, callback) {
+    let files = [];
+    let clientFiles = [];
+    let serverFiles = [];
+    let testFiles = [];
+    let index = 0;
+
+    const ignoreHidden = item => {
+        const basename = path.basename(item);
+        return basename === '.' || basename[0] !== '.';
+    };
+
+    const walker = klaw(options.baseDir, {filter: ignoreHidden});
+
+    options.ignoreDirs = options.ignoreDirs || [];
+
+    walker.on('readable', () => {
+        for (;;) {
+            const item = walker.read();
+            if (!item) {
+                break;
+            }
+            if (!item.stats.isDirectory()) {
+                const relPath = path.relative(options.baseDir, item.path);
+                const prefix = relPath.split(path.sep)[0];
+                const file = {
+                    name: relPath,
+                    path: path.relative(options.courseDir, item.path),
+                    editable: canEditFile(item.path),
+                    moveable: canMoveFile(item.path),
+                    id: index,
+                };
+                if (prefix == options.clientFilesDir) {
+                    clientFiles.push(file);
+                    index++;
+                } else if (prefix == options.serverFilesDir) {
+                    serverFiles.push(file);
+                    index++;
+                } else if ((options.testFilesDir) && (prefix == options.testFilesDir)) {
+                    testFiles.push(file);
+                    index++;
+                } else if (! options.ignoreDirs.includes(prefix)) {
+                    files.push(file);
+                    index++;
+                }
+            }
+        }
+    });
+
+    walker.on('error', (err) => {
+        if (ERR(err, callback)) return;
+    });
+
+    walker.on('end', () => {
+        callback(null, {
+            files: files,
+            clientFiles: clientFiles,
+            serverFiles: serverFiles,
+            testFiles: testFiles,
+            dir: path.relative(options.courseDir, options.baseDir),
+        });
+    });
+}
+
+function getHashFromBuffer(buffer) {
+    return sha256(buffer.toString('utf8')).toString();
+}
+
+function file_delete_write(edit, callback) {
+    debug(`Delete ${edit.filePath}`);
+    // This will silently do nothing if edit.filePath no longer exists.
+    fs.remove(edit.filePath, (err) => {
+        if (ERR(err, callback)) return;
+        edit.pathsToAdd = [
+            edit.filePath,
+        ];
+        edit.commitMessage = `in-browser edit: delete file ${edit.filePath}`;
```

Comment:
Would it be difficult to have commit messages on an "object" level, rather than a "file" level? That is, indicate that we are editing a specific question or course instance, rather than focus on the file that was edited?

To keep commit messages as short as possible, I'm thinking about something like: `<QID>: delete <filePath>`, where `<filePath>` is relative to the question. This would give us messages like:
```
Fall2019: edit infoCourseInstance.json
Fall2019: delete clientFilesCourseInstance/formulas.pdf
demoCalculation: edit server.py
demoCalculation: edit question.html
```

Reply:
Done.